lighthouse-core/report/html/renderer/code-snippet-renderer.js

```diff
@@ -0,0 +1,283 @@
+      number: ' ',
+      content: highlight.message,
+      contentType: LineContentType.MESSAGE,
+    });
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {DocumentFragment} tmpl
+   * @param {LineVisibility} visibility
+   * @return {Element}
+   */
+  static renderOmittedLines(dom, tmpl, visibility) {
+    return CodeSnippetRenderer.renderSnippetLine(dom, tmpl, {
+      number: '…',
+      content: '',
+      visibility,
+      contentType: LineContentType.EMPTY,
+    });
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {DocumentFragment} tmpl
+   * @param {LH.Audit.DetailsRendererCodeSnippet} details
+   * @return {DocumentFragment}
+   */
+  static renderSnippet(dom, tmpl, details) {
+    const {highlights, lineCount, lines} = details;
+
+    const template = dom.cloneTemplate('#tmpl-lh-code-snippet__content', tmpl);
+    const snippet = dom.find('.lh-code-snippet__snippet-inner', template);
+
+    const nonLineSpecificHighlights = highlights.filter(h => typeof h.lineNumber !== 'number');
+    nonLineSpecificHighlights.forEach(h =>
+      snippet.append(CodeSnippetRenderer.renderHighlightMessage(dom, tmpl, h))
+    );
+
+    snippet.append(CodeSnippetRenderer.renderSnippetLines(dom, tmpl, details));
+
+    // If expanded view still doesn't include all lines then indicate that that
+    const firstLineIsVisible = lines[0].number === 1;
+    const lastLineIsVisible = lines[lines.length - 1].number === lineCount;
+    if (!firstLineIsVisible) {
+      snippet.prepend(CodeSnippetRenderer.renderOmittedLines(dom, tmpl,
+         LineVisibility.WHEN_EXPANDED
+      ));
+    }
+    if (!lastLineIsVisible) {
+      snippet.append(CodeSnippetRenderer.renderOmittedLines(dom, tmpl,
+        LineVisibility.WHEN_EXPANDED
+      ));
+    }
+
+    return template;
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {DocumentFragment} tmpl
+   * @param {LH.Audit.DetailsRendererCodeSnippet} details
+   * @returns {DocumentFragment}
+   */
+  static renderSnippetLines(dom, tmpl, details) {
+    const {highlights, lineCount, lines} = details;
+    const collapsedLines = Util.filterRelevantLines(lines, highlights, 2);
+    const hasLineSpecificHighlights = highlights.some(h => typeof h.lineNumber === 'number');
+    const hasOnlyNonLineSpecficHighlights = highlights.length > 0 && !hasLineSpecificHighlights;
+
+    const lineContainer = dom.createFragment();
+
+    // Keep track of whether a highlighted line has been seen, because we don't want this code
+    // to render an omitted lines indication before the first line
+    let hasSeenHighlight = false;
+    for (let lineNumber = 1; lineNumber <= lineCount; lineNumber++) {
+      const {line, previousLine} = getLineAndPreviousLine(lines, lineNumber);
+      const {line: collapsedLine, previousLine: collapsedPreviousLine} =
+        getLineAndPreviousLine(collapsedLines, lineNumber);
+
+      if (hasSeenHighlight) {
+        // Show if some lines were omitted
+        if (line && !previousLine) {
```

Comment:
is it true that `collapsedLine` can never be truthy if `line` is not truthy? 
is it the same for `previousLine` and `collapsedPreviousLine`?

what I'm getting at here is that it feels like we have 2 cases of omitted line placeholders not 3, either this line is always shown as omitted or it's only shown as omitted when collapsed.

Reply:
Here's an example, (X) means only in expanded lines array, `-` means the line is not in either array

```
 A 
(B) => should insert "..." before this, but only if collapsed because
            there are no missing lines before C
 C
(D) => hmm, actually this is a bit trickier... when we want to show "..."
            for both we want to show it after the last line that's visible in
            the expanded state, so we need to look forward when the 
            line is missing in linesVisibleWhenCollapsed, and backward
            if the line is missing in all available lines
 -
(K)
 L
(M)
```

Reply:
I guess it can all just be handled by setting flags, so it's less complex than I thought.